ckanext/external_storage/helpers.py

```diff
@@ -19,15 +22,27 @@ def resource_storage_prefix(package_name, org_name=None):
     return '{}/{}'.format(org_name, package_name)
 
 
-def resource_authz_scope(package_name, actions=None, org_name=None, resource_id=None):
+def resource_authz_scope(package_name, actions=None, org_name=None, resource_id=None, activity_id=None):
```

Comment:
That was the hardest obstacle while trying to support of downloading resource files from arbitrary activity_id.
The serialized scope object is always created for a `resource_id`. This works good for checking the access permissions.
Only the `resource_id` is later used to determine `sha256` key to be fetched from giftless.

The only way to bypass that I could think of is to create a composite `resource_id/activity_id` key.